src/main/groovy/de/dkfz/roddy/bam2fastq/BamToFastqWorkflow.groovy

```diff
@@ -0,0 +1,266 @@
+        if (Roddy.isMetadataCLOptionSet()) {
+            logger.severe("Metadata table input not implemented. Please use ${Config.CVALUE_BAMFILE_LIST} to specify the BAM files to convert.")
+        } else if (config.bamList.size() > 0) {
+            bamFiles = config.bamList.collect { filename ->
+                BaseFile.getSourceFile(context, filename, "BamFile")
+            }
+        } else {
+            // Collect files from directory structure.
+            logger.severe("Please use '${Config.CVALUE_BAMFILE_LIST}' to specify the BAM files to convert.")
+        }
+        if (bamFiles.size() == 0)
+            logger.warning("No input BAM files were specified for dataset ${context.dataSet}.")
+
+        _bamFilesPerDataset[context.dataSet] = bamFiles
+    }
+
+    /**
+     * Return a list of read-groups identifiers from a bam
+     * @param context
+     * @param bamfileName
+     */
+    private synchronized final static Map<String,List<String>> readGroupsPerBamfile = [:]
+
+    List<String> listReadGroups(String bamfileName) {
+        if (!readGroupsPerBamfile[bamfileName]) {
+            logger.always("Listing read groups in '$bamfileName'.")
+            readGroupsPerBamfile[bamfileName] =
+                    ExecutionService.getInstance().runDirect(context, TOOL_BAM_LIST_READ_GROUPS, ["BAMFILE": bamfileName] as Map<String, Object>)
+        }
+        return readGroupsPerBamfile[bamfileName]
+    }
+
+    static List<String> readGroupIndices(List<String> readGroups) {
+        return readGroups.collect { String rg -> [rg + "_R1", rg + "_R2"] }.flatten() as List<String>
+    }
+
+    /**
+     * Extract FASTQs from a BAM. This may be with or without read-groups. The names of the output files depend on the parameters.
+     */
+    FileGroup bam2fastq(BaseFile controlBam, List<String> readGroups) {
+        if (config.pairedEnd) {
+            List<String> rgFileIndicesParameter = readGroupIndices(readGroups)
+            return callWithOutputFileGroup(TOOL_BAM2FASTQ, controlBam, rgFileIndicesParameter, [readGroups: readGroups])
+        } else {
+            throw new ConfigurationError("Single-end bam2fastq not implemented", config.FLAG_PAIRED_END)
+        }
+    }
+
+    /** Submit a sorting job for each read group. The read-group name is provided via the `readGroup` parameter. Also the BAM-file name is provided
+     *  such that it can be used in the filename pattern using the `${cvalue}` syntax.
+     *
+     * @param bamFileName
+     * @param readGroup
+     * @param fastq1
+     * @param fastq2
+     * @param fastq3
+     */
+    void sortFastqs(String bamFileName, String readGroup, BaseFile fastq1, BaseFile fastq2 = null, BaseFile fastq3 = null) {
+        assert(fastq1 != fastq2)
+        assert(fastq1 != fastq3)
+        assert(fastq2 == null || fastq2 != fastq3)
+        HashMap<String, String> parameters = [readGroup: readGroup, bamFileName: bamFileName]
+        if (config.pairedEnd) {
```

Comment:
With all these if's, it might be easier to have one base and two subclasses for the workflow.
```
      BaseBam2FastqWorkflow
          /            \
       PairedEnd     SingleEnd (Not implemented)
```

Reply:
Also it could be much easier to break before all this. We have the checkexecutability and you could just check for single end there. Then you could leave the elses blank here.